Pharmacy/PharamcyApi/Controllers/MedicineController.cs

```diff
@@ -102,6 +94,65 @@ public IActionResult GetById(int id)
             return Ok(medicine);
         }
 
+        [HttpGet]
+        public IActionResult GetFilteredMedicine(string medicineName, string substanceName, string medicineType, string manufacturerName)
+        {
+            IEnumerable<Medicine> medicines = _uow.GetRepository<IMedicineReadRepository>().GetAll()
+                .Include(medicine => medicine.Manufacturer)
+                .Include(medicine => medicine.Substances)
+                .Where(medicine => String.IsNullOrEmpty(medicineName) || medicine.Name.Equals(medicineName))
+                .Where(medicine => String.IsNullOrEmpty(substanceName) || (medicine.Substances.Where(substance => substance.Name == substanceName).Any()))
+                .Where(medicine => String.IsNullOrEmpty(medicineType) || medicine.Type.Equals(medicineType))
+                .Where(medicine => String.IsNullOrEmpty(manufacturerName) || medicine.Manufacturer.Name.Equals(manufacturerName));
+
+            if (medicines == null || medicines.Count() == 0)
+                return NotFound();
+
+            return Ok(medicines);
+        }
+
+        [HttpGet]
+        public IActionResult GetFilteredMedicineWithPaging(string medicineName, string substanceName, string medicineType, string manufacturerName, int pageNumber, int pageSize)
```

Comment:
Maybe instead of multiple parameters use somekind of wrapper object like FilteredMedicineDTO where you could group them.